storey/timescaledb_target.py

```diff
@@ -0,0 +1,304 @@
+        ...     table="sensor_data",
+        ...     time_format="%Y-%m-%d %H:%M:%S.%f",
+        ...     max_events=1000,
+        ...     flush_after_seconds=5
+        ... )
+
+        >>> # Usage with event metadata
+        >>> target = TimescaleDBTarget(
+        ...     dsn="postgresql://user:pass@localhost:5432/mydb",
+        ...     time_col="event_time",
+        ...     columns=["$key", "value", "source=device_name"],
+        ...     table="events"
+        ... )
+
+    Note:
+        - The target table must be created as a TimescaleDB hypertable before use
+        - The time column should be a timestamp type, preferably TIMESTAMPTZ for timezone awareness
+        - Events are written using PostgreSQL's COPY protocol for optimal performance
+        - Connection pooling is handled automatically with proper cleanup on termination
+    """
+
+    def __init__(
+        self,
+        dsn: str,
+        time_col: str,
+        columns: list[str],
+        time_format: Optional[str] = None,
+        table: Optional[str] = None,
+        max_connections: int = 1,
+        min_connections: int = 1,
+        **kwargs,
+    ) -> None:
+
+        # Pass parameters to parent classes
+        kwargs["time_col"] = time_col
+        kwargs["columns"] = columns
+        if table:
+            kwargs["table"] = table
+        if time_format:
+            kwargs["time_format"] = time_format
+
+        # Initialize batching functionality
+        _Batching.__init__(self, **kwargs)
+
+        # Initialize writer functionality with time column as first column
+        # This ensures proper column ordering for TimescaleDB's time-partitioned structure
+        _Writer.__init__(
+            self,
+            columns=[time_col] + columns,
+            infer_columns_from_data=False,
+            retain_dict=True,
+            time_field=time_col,
+            time_format=time_format,
+        )
+        self._table = table
+
+        # Store configuration
+        self._time_col = time_col
+        self._columns = columns
+
+        # Database connection configuration
+        self._dsn = dsn
+        self._max_connections = max_connections
+        self._min_connections = min_connections
+        self._pool = None  # Connection pool will be created lazily during first use
+        self._column_names = self._get_column_names()
+
+    def _test_connection_sync(self) -> None:
+        """Test database connection synchronously during initialization.
+
+        Only tests connection if not in an async context.
+        """
+        with contextlib.suppress(RuntimeError):
+            # Check if we're in an async context
+            asyncio.get_running_loop()
+            # If we get here, we're in an async context - skip sync testing
+            return
+        try:
+            loop = asyncio.new_event_loop()
+            asyncio.set_event_loop(loop)
+            loop.run_until_complete(self._test_connection_async())
+            loop.close()
```

Comment:
Not really sure if this validation is really necessary in `_init()` at all, but if it is, it can be simplified:
```suggestion
            asyncio.run(self._test_connection_async())
```

Reply:
The verification is done on calling enable_model_monitoring() - this is a sync call and it just checks the validity of the credentials and the connection. Later on its not used. Changing as per your suggestion

Reply:
I don't think this is correct though. Calling `enable_model_monitoring()` will deploy MM functions, yes, but these functions run in nuclio, which means they run in an async context. So the check will be skipped. As I can tell, this code is going to be no-op.